Rationalize notifications

atoghill opened the pull request:
Resolves #507. Notifications should now follow the table provided within that issue:

| Event                                 | Email: site contacts | Email: support team | Slack |
|---------------------------------------|----------------------|---------------------|-------|
| New case (by site contact)            | ✓                    | ✓                   | ✓     |
| New case (by admin)                   | ✓                    | ✓                   | ✓     |
| Change case engineer ("assignee")     |                      | ✓                   | ✓     |
| Change case subject                   | ✓                    | ✓                   | ✓     |
| Change case issue                     |                      | ✓                   | ✓     |
| Change case associations              |                      | ✓                   | ✓     |
| Add case comment                      | ✓                    | ✓                   | ✓     |
| Escalate support tier                 |                      |                     |       |
| Set case time worked                  |                      |                     |       |
| Case resolution                       |                      | ✓                   | ✓     |
| Case closure                          |                      |                     |       |
| Maintenance window state transition   | ✓                    | ✓                   | ✓     |
| Maintenance ending soon               |                      | ✓                   | ✓     |
| Change request: draft creation        |                      | ✓                   | ✓     |
| Change request: customer state change | ✓                    | ✓                   | ✓     |
| Change request: draft cancellation    |                      | ✓                   | ✓     |
| Log entry creation                    |                      | ✓                   | ✓     |
| Daily check completion                |                      | ✓                   | ✓     |

Minor changes could be made to clean things up but the core changes should all be present in this PR.

[Trello](https://trello.com/c/Jwnw3bvv/442-rationalize-notifications)

jamesremuscat commented on spec/support/shared_examples/slack_notifications.rb:
```diff
@@ -0,0 +1,8 @@
+RSpec.shared_examples 'Slack' do
+  it 'sends a notification to Slack' do
+    expect(SlackNotifier).to receive(notification_method)
+      .with(*args)
```
Could we rename `args` to `slack_args`? It's often used some way away from any reference to Slack and could do with being a bit more descriptive.

jamesremuscat commented on app/mailers/admin_mailer.rb:
```diff
@@ -10,4 +11,22 @@ def daily_open_cases_list(admin, cases)
     )
   end
 
+  def new_log(log)
```
Please could you add a corresponding method in `admin_mailer_preview'?

jamesremuscat commented on app/mailers/admin_mailer.rb:
```diff
@@ -10,4 +11,22 @@ def daily_open_cases_list(admin, cases)
+    SlackNotifier.log_notification(@log, @text)
+  end
+
+  def cluster_check_submission(cluster, user)
```
Please could you add a corresponding method in `admin_mailer_preview'?

jamesremuscat commented on app/mailers/case_mailer.rb:
```diff
@@ -79,16 +80,38 @@ def maintenance_ending_soon(window, text)
     SlackNotifier.change_request_notification(@case, @text, user)
   end
 
+  def change_association(my_case, user)
```
Please could you add a corresponding method in `case_mailer_preview'?

jamesremuscat commented on app/mailers/case_mailer.rb:
```diff
@@ -79,16 +80,38 @@ def maintenance_ending_soon(window, text)
+    SlackNotifier.case_association_notification(@case, user, @text)
+  end
+
+  def resolve_case(kase, user)
```
Please could you add a corresponding method in `case_mailer_preview'?

atoghill commented:
@jamesremuscat I have addressed all your comments now. In regards to the emails themselves I was unsure what and how much content we wanted within them so do let me know if they need adjusting

jamesremuscat commented on app/mailers/case_mailer.rb:
```diff
@@ -79,16 +80,38 @@ def maintenance_ending_soon(window, text)
+    @case = my_case
+    reference_texts = @case.associations
+      .map { |a| a.decorate.reference_text }
+    @text = %{Changed the affected components on this case to:
```
This text doesn't really work if there are no components associated with a case (which is a valid state).

atoghill replied:
It is a valid state for a case but is there ever a scenario where this email would be sent in such a case? As far as I can tell once there are components associated with a case you can't completely remove all associations via the main site. Unless I am misunderstanding what you mean.

jamesremuscat replied:
Your "telling" is wrong. It's entirely possible to remove all associations from a case, so long as the associated `Issue` doesn't require them.
